### Encoding

In [26]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import OneHotEncoder, LabelEncoder

In [3]:
# creating sample data
sample_data = {'name': ['Ray', 'Adam', 'Jason', 'Varun', 'Xiao'],
'health':['fit', 'slim', 'obese', 'fit', 'slim']}
# storing sample data in the form of a dataframe
data = pd.DataFrame(sample_data, columns = ['name', 'health'])

In [4]:
data

name health
0    Ray    fit
1   Adam   slim
2  Jason  obese
3  Varun    fit
4   Xiao   slim

In [6]:
label_encoder = LabelEncoder()
label_encoder.fit_transform(data['health'])

array([0, 2, 1, 0, 2])

In [9]:
pd.get_dummies(data['health'])

fit  obese  slim
0    1      0     0
1    0      0     1
2    0      1     0
3    1      0     0
4    0      0     1

In [25]:
labels = label_encoder.fit_transform(data['health'])
one_hot_encoder = OneHotEncoder()
one_hot_encoder.fit_transform(labels.reshape(-1,1)).toarray()

array([[ 1.,  0.,  0.],
       [ 0.,  0.,  1.],
       [ 0.,  1.,  0.],
       [ 1.,  0.,  0.],
       [ 0.,  0.,  1.]])

In [37]:
# Load the dataset
X = pd.read_csv('titanic_data.csv')
# Limit to categorical data
X = X.select_dtypes(include=[object])

# TODO: Create a LabelEncoder object, which will turn all labels present in
#       in each feature to numbers. 
# HINT: Use LabelEncoder()
le = LabelEncoder()

# TODO: For each feature in X, apply the LabelEncoder's fit_transform
#       function, which will first learn the labels for the feature (fit)
#       and then change the labels to numbers (transform). 

for feature in X:    
    X[feature] = le.fit_transform(X[feature])

# TODO: Create a OneHotEncoder object, which will create a feature for each
#       label present in the data. 
# HINT: Use OneHotEncoder()
ohe = OneHotEncoder()

# TODO: Apply the OneHotEncoder's fit_transform function to all of X, which will
#       first learn of all the (now numerical) labels in the data (fit), and then
#       change the data to one-hot encoded entries (transform).

# HINT: Use fit_transform on X using the OneHotEncoder() object
xt = ohe.fit_transform(X)
onehotlabels = xt


In [42]:
onehotlabels.toarray().shape

(891, 1726)

In [51]:
XX = pd.read_csv('titanic_data.csv')
XX = XX.select_dtypes(include=[object])
XX.head()

Name     Sex  \
0                            Braund, Mr. Owen Harris    male   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2                             Heikkinen, Miss. Laina  female   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4                           Allen, Mr. William Henry    male   

             Ticket Cabin Embarked  
0         A/5 21171   NaN        S  
1          PC 17599   C85        C  
2  STON/O2. 3101282   NaN        S  
3            113803  C123        S  
4            373450   NaN        S

In [52]:
print 'Unique values per column:', ohe.n_values_

Unique values per column: [891   2 681 148   4]


In [50]:
print 'Encoded classes\' positions:', ohe.feature_indices_

Encoded classes' positions: [   0  891  893 1574 1722 1726]


### name has 891 values encoded in position 0-890, sex has 2 values encoded in 891-892